### PR TITLE
Name expressions after typing in ocamlnat (restores equivalence with bytecode toplevel)

### PR DESCRIPTION
**Pull request by NathanReb**
## Description of the bug

Running the following code was raising different results in bytecode and native toplevels:
```ocaml
module A : sig 
  type ('foo, 'bar) t

  val get_foo : ('foo, _) t -> 'foo option
end = struct
  type ('foo, 'bar) t =
    | Foo of 'foo
    | Bar of 'bar

  let get_foo = function
    | Foo foo -> Some foo
    | Bar _ -> None
end
;;

A.get_foo
;;
```

In the bytecode toplevel we get the following:
```ocaml
- : ('foo, 'a) A.t -> 'foo option = <fun>
```
which is what we would expect.

In the native toplevel though, the type variable name is lost and we instead get the following:
```ocaml
- : ('a, 'b) A.t -> 'a option = <fun>
```

## The fix

This was written by @lpw25 . It makes the native toplevel preserve type variable names in the same way as the bytecode toplevel.

This patch was originally submitted to flambda-backend by @mshinwell here: https://github.com/ocaml-flambda/flambda-backend/pull/188.

### CODE REVIEW

**Comment by NathanReb** [comment]
CC @mshinwell @dra27

**Comment by gasche** [comment]
Apologies for the naive question, but: why is this part of the `topeval.ml` codebase so different between the `byte` and the `native` versions? The code is relatively different today, and the new patch proposed today adds a new auxiliary function in `native` that does not seem to correspond to the logic in the bytecode toplevel.

**Comment by NathanReb** [comment]
To be entirely honest with you I don't know for certain. As you point out the code differs quite a bit between the native and bytecode toplevels already here. It might be the case that the native toplevel would simply need to be brought up to speed with the bytecode on this particular thing. This fix was written on top of 4.12 which doesn't have the unified toplevels.

This PR does fix the bug but if you're not willing to merge it as it is I can try to look into this further and see if this can be shared across bytecode and native.

**Comment by NathanReb** [comment]
Just rebased and fixed the `.depend` files!

**Comment by trefis** [comment]
I'm sorry but I don't understand how that patch has the effect you claim it has.
From what I understand, previously we were rewriting the parsetree to name the expression before calling the typechecker, now we rewritte the typedtree after calling the typechecker.

The way things are typechecked should be virtually unchanged, the same goes for the way things are printed.
The only meaningful change is that before you add a weird "_$" value inserted in the toplevel environment, and now you don't anymore.

What am I missing?

**Comment by lpw25** [comment]
IIRC the issue is that we get different output typing `let foo = e` than typing just `e` as a toplevel expression. This patch fixes it by always typing the code as it was written and then doing a transformation on the typed AST rather than the untyped AST.

I think it's also a little cleaner than the old approach because it makes no assumptions about the shape of the signature produced by the type-checker. Instead it directly produces a signature of the shape that the later code expects when `rewritten` is true.

**Comment by trefis** [comment]
> IIRC the issue is that we get different output typing `let foo = e` than typing just `e` as a toplevel expression

Again, I failed to understand how/why yesterday.
But I can't contest that it is true, here in the **bytecode** toplevel:

```
# A.get_foo;;
- : ('foo, 'a) A.t -> 'foo option = <fun>
# let foo = A.get_foo;;
val foo : ('a, 'b) A.t -> 'a option = <fun>
```

I think at this point my opinion of the PR is the following: it is *not* fixing the bug it claims to fix; it's doing some "things", which happen to make the printing of type variables by the native toplevel a bit nicer, in some cases. The "things" it's doing might be things we want to do, but it's also increasing the distance between the bytecode and native toplevels.

What I'd like to see is the issue of type variable printing considered properly: explaining how and why it's there, and what the proper fix should be (it might very well be that we have to change things in the toplevels, but *a priori* that sounds surprising).
Independently, I'd welcome a PR which changes *consistently* the way the toplevel handles expressions which aren't bound.

**Comment by gasche** [comment]
I concur on one point: if the change is about typing (or printing the types, etc.), I would expect the final code to be the same in the bytecode and native toplevel.

(I understand that this may require some preparation work to understand and avoid the existing differences between the bytecode and native toplevel. But it's much better to tackle the question now rather than later, after you have changed more things in one of the toplevels. The goal, if I understand correctly, is to end in a state where we improve and maintain the two toplevels, and changing only one side adds more technical debt.)

**Comment by lpw25** [comment]
This PR does make the code in the native version closer to the code in the bytecode version. The parts that aren't the same are directly due to how the native code versions executes its terms.

**Comment by lpw25** [comment]
I think this discussion, and I guess the PR description, are somewhat off-topic. The bug here is that the native and bytecode toplevels do not agree on what term should be type-checked. This can lead to differences in the output of the type-checker. 

The fact that different but semantically equivalent terms can produce different output from the type-checker is not a bug, rather it is to be expected. The bug is that the toplevels have diverged in how they interact with the type-checker. This PR directly fixes that problem by getting the two toplevels to agree.

**Comment by dra27** [comment]
I took the liberty of rebasing this, while adding a test. @shindere - does adding `toplevel.opt` seem OK, here?

**Comment by dra27** [comment]
This PR seems to have more bugs in the discussion than in the code 🙂 I've attempted to mitigate this by updating both the PR title and Changes to hopefully reflect a little better what is wrong and why this fixes it!
- ocamlnat necessarily has to name expressions in order to load them (at a high level, that's just natdynlink going on in the background). ocaml doesn't have to do that because virtual machine.
- @lpw25's fix moving the "transformation" from parse tree to typed tree looks ok to me - the difference in (code) size may at least partly explain why it wasn't that way in the first place. The two toplevels do fairly clearly now type-check the same expression.
- I agree that the behaviour seen in the test is a _bug_. We would not expect ocamlc and ocamlopt to produce anything other than identical types for the same piece of code; it seems reasonable for a user to expect that ocaml and ocamlnat will do the same for an identical phrase entered in the toploop.
- I braved the pandemic to beg @Octachron in-person for this fix's inclusion in 4.14 in January, so I've added it to the milestone, but at least he should approve that before it's merged (and, of course, the previous dissenting opinions still matter!)

**Comment by gasche** [comment]
> ocamlnat necessarily has to name expressions in order to load them (at a high level, that's just natdynlink going on in the background). ocaml doesn't have to do that because virtual machine.

That makes sense, and indeed it was the big missing piece.

> @lpw25's fix moving the "transformation" from parse tree to typed tree looks ok to me

okay, so given this binding-alteration step, what's going on makes sense. It is necessary to transform the program to name all expressions. The previous transformation would result in changes in type-checking behavior. The new transformation happens on the typed-tree, so it does not risk changing inferred types anymore.

This explanation was entirely missing from the discussion, or the previous code, or the new code, or the commit message, or the description of the two PRs. Could you make sure that the relevant aspects are documented somewhere in native/topeval.ml as a comment?

> I braved the pandemic to beg @Octachron in-person for this fix's inclusion in 4.14 in January, so I've added it to the milestone, but at least he should approve that before it's merged

I guess it's going to be easier to review the patch with a proper explanation of what is going on, thanks!

>  the difference in (code) size may at least partly explain why it wasn't that way in the first place

I thought at first that you meant that this logic increases the size of the code generated by the toplevel. But in fact you mean that the patch uses a longer implementation, right?

> (and, of course, the previous dissenting opinions still matter!)

I looked at the code again to understand whether the bytecode and toplevel logic could be made closer to each other. The difference between the two now makes sense, so I would be happy to approve it *if* someone clearly documents what this code is doing and why it is doing it that way.

**Comment by gasche** [comment]
Note: the i386-static CI build seems to fail with an issue relevant to the PR.

```diff

+++ /home/runner/work/ocaml/ocaml/testsuite/_ocamltest/tests/tool-toplevel/pr10712/ocamlnat/ocamlnat.output	2021-12-20 19:39:38.368908459 +0000
@@ -1,5 +1,12 @@
-module A :
-  sig type ('foo, 'bar) t val get_foo : ('foo, 'a) t -> 'foo option end
-- : ('foo, 'a) A.t -> 'foo option = <fun>
-val _bar : ('a, 'b) A.t -> 'a option = <fun>
+sh: 1: shared-libs-not-available: not found
+Line 1:
+Error: I/O error: shared-libs-not-available -o 'TOP1.so'  '-L/home/runner/work/ocaml/ocaml/stdlib' '-L/home/runner/work/ocaml/ocaml/toplevel'  'TOP1.o' 
+Line 3, characters 0-9:
+3 | A.get_foo
+    ^^^^^^^^^
+Error: Unbound module A
+Line 3, characters 11-20:
+3 | let _bar = A.get_foo
+               ^^^^^^^^^
+Error: Unbound module A
```

**Comment by dra27** [comment]
> > the difference in (code) size may at least partly explain why it wasn't that way in the first place
> 
> I thought at first that you meant that this logic increases the size of the code generated by the toplevel. But in fact you mean that the patch uses a longer implementation, right?

That's correct, yes - I just meant that it requires slightly longer sources to manipulate it at the type tree level, so in the absence of a reason (such as the type-checker behaving slightly differently) it may have been tempting to alter it at the parse tree level before.

**Comment by dra27** [comment]
CI hopefully correctly fixed - it's correct that `ocamltoplevel.cmxa` is compiled even in non-shared mode because the point of the new hooks is that you may add shared library support another way, but obviously the testsuite should not attempt to run `ocamlnat` in that configuration.

**Comment by gasche** [comment]
Re. "difference between the bytecode and native versions": I think some well-targeted refactoring could improve the similarity between the two codebases by extracting the differences into helper functions. But this is not strictly required here, and can be done in a later PR.

Re. "backporting in 4.14": I suppose you were convincing in-person, but could you say a few words about why we care to have this urgently in 4.14? From a distance it does not look like an urgent fix, but surely there is a good explanation.

**Comment by dra27** [comment]
Yes, indeed - the "urgency" if you like is that we managed to get the hooks and so forth into the toplevel to allow the ocaml-jit project to work in the 4.x series without needing a patched compiler, but the main use of it at the moment is for tools such as [mdx](https://github.com/realworldocaml/mdx). The feature is much less useful if the bytecode and native toplevels are so easily able to produce different output (which is how the bug was found). From a "safety" perspective, choosing to fix this in 4.14 just before the first alpha/beta, 4.14 is also the first release where we officially install the native toplevel library at all.

**Comment by dra27** [comment]
I was only convincing inasmuch as @Octachron didn't say "absolutely, no" 🙂

**Comment by Octachron** [comment]
To detail my opinion, since ocamlnat is still only enabled by an explicit configuration option, I think that it sensible to integrate in 4.14/5.0 bug fixes that will make the ongoing work on ocamlnat easier in the next development cycle.

**Comment by gasche** [comment]
Current status: I'm waiting for some documentation comments to explain what's going on, and then I would be willing to approve the PR.

**Comment by dra27** [comment]
Oops - I'd managed to miss that as being a "TODO" item. Comment added!

**Review by gasche — APPROVED** [review]
I like the new documentation, thanks! Approved.

(You could squash the two documentation commits.)

**Comment by dra27** [comment]
Squashed and tidied. thanks!

**Comment by shindere** [comment]
David Allsopp (2021/12/20 11:28 -0800):
> I took the liberty of rebasing this, while adding a test. @shindere -
> does adding `toplevel.opt` seem OK, here?

Coming in late but, yes. Sorry for the delay.

**Pull request merged**
